3d_armor/init.lua

```diff
@@ -394,6 +394,9 @@ end
 if armor.config.punch_damage == true then
 	minetest.register_on_punchplayer(function(player, hitter,
 			time_from_last_punch, tool_capabilities)
+		if hitter == nil then
+			return
```

Comment:
Thanks for addressing this. Are you sure this is the correct course of action, though? Wouldn't it make more sense to pass the `nil` hitter on to `armor:punch` (which then needs to support `hitter == nil`, which it already seems to do)?

This early return defers the damage handling to `on_player_hpchange` where the `tool_capabilities` - and hence the "type" of damage - is unknown if I'm not mistaken. So for example if a mod wanted to deal "radiation" damage without a puncher, this would early return, and the damage would be handled in `on_player_hpchange`, where it would not be known that it is "radiation" damage anymore.

I think a better fix might be `local hit_ip = hitter and hitter:is_player()`, assuming `armor:punch` supports a `nil` `hitter`. What do you think?

Reply:
Changed in https://sprunge.us/GESjsB?diff. I added an extra `nil` check when `armor:punch` calls custom armor `def.on_punched` functions since they may require `puncher ~= nil`

Reply:
Zipgrep is here: https://web.archive.org/web/20240318145231/https://content.minetest.net/zipgrep/14aae5e3-1350-4db7-baa2-d127abd0f021/

`on_punched` is only used in `halloween` (doesn't really do anything, will work) and `aliveai` (I'm pretty sure [this code will break](https://github.com/AiTechEye/aliveai/blob/master/aliveai_threats/init.lua)).

We should propagate `nil` punchers to custom functions IMO.

Reply:
The `aliveai` code seems unrelated to `3d_armor`, though. I think it's just coincidentally named the same as the `3d_armor` callback.

It looks like it's safe to propagate `nil` punchers :+1: